aiosmtpd/smtp.py

```diff
@@ -865,42 +902,75 @@ async def smtp_DATA(self, arg: str) -> None:
         if arg:
             await self.push('501 Syntax: DATA')
             return
+
         await self.push('354 End data with <CR><LF>.<CR><LF>')
-        data = []
-        num_bytes = 0
-        size_exceeded = False
+        data: List[bytes] = []
+
+        num_bytes: int = 0
+        limit: Optional[int] = self.data_size_limit
+        partial_line: List[bytes] = []
+        state: _DataState = _DataState.NOMINAL
         while self.transport is not None:           # pragma: nobranch
+            # Since eof_received cancels this coroutine,
+            # readuntil() can never raise asyncio.IncompleteReadError.
             try:
-                line = await self._reader.readline()
+                line: bytes = await self._reader.readuntil()
                 log.debug('DATA readline: %s', line)
+                assert line.endswith(b'\n')
             except asyncio.CancelledError:
                 # The connection got reset during the DATA command.
                 log.info('Connection lost during DATA')
                 self._writer.close()
                 raise
-            if line == b'.\r\n':
+            except asyncio.LimitOverrunError as e:
+                # The line exceeds StreamReader's buffer.
+                # Delay SMTP Status Code sending until data receive is complete
+                # This seems to be implied in RFC 5321 § 4.2.5
+                state = _DataState.TOO_LONG
+                # Discard data immediately to prevent memory pressure
+                data *= 0
```

Comment:
Did some tests, and `*= 0` is indeed faster by about 10-20% compared to `.clear()` or `del [:]`. And it's _still_ in-place. Go figure.